lib/obj.js

```diff
@@ -1,365 +1,479 @@
+            });
+    }).then(function() {
+        return images;
+    });
+}
+
+function getImagePaths(materials) {
+    var imagePaths = {};
     for (var name in materials) {
         if (materials.hasOwnProperty(name)) {
             var material = materials[name];
-            if (defined(material.ambientColorMap) && (images.indexOf(material.ambientColorMap) === -1)) {
-                images.push(material.ambientColorMap);
+            if (defined(material.ambientColorMap)) {
+                imagePaths[material.ambientColorMap] = true;
             }
-            if (defined(material.diffuseColorMap) && (images.indexOf(material.diffuseColorMap) === -1)) {
-                images.push(material.diffuseColorMap);
+            if (defined(material.diffuseColorMap)) {
+                imagePaths[material.diffuseColorMap] = true;
             }
-            if (defined(material.emissionColorMap) && (images.indexOf(material.emissionColorMap) === -1)) {
-                images.push(material.emissionColorMap);
+            if (defined(material.emissionColorMap)) {
+                imagePaths[material.emissionColorMap] = true;
             }
-            if (defined(material.specularColorMap) && (images.indexOf(material.specularColorMap) === -1)) {
-                images.push(material.specularColorMap);
+            if (defined(material.specularColorMap)) {
+                imagePaths[material.specularColorMap] = true;
             }
         }
     }
+    return Object.keys(imagePaths);
+}
 
-    // Load the image files
-    var promises = [];
-    var imagesInfo = {};
-    var imagesLength = images.length;
-    for (var i = 0; i < imagesLength; i++) {
-        var imagePath = images[i];
-        if (!path.isAbsolute(imagePath)) {
-            imagePath = path.join(inputPath, imagePath);
+function removeEmptyPrimitives(primitives) {
+    var final = [];
+    var primitivesLength = primitives.length;
+    for (var i = 0; i < primitivesLength; ++i) {
+        var primitive = primitives[i];
+        if (primitive.indices.length > 0) {
+            final.push(primitive);
         }
-        promises.push(loadImage(imagePath));
     }
-    return Promise.all(promises)
-        .then(function(imageInfoArray) {
-            var imageInfoArrayLength = imageInfoArray.length;
-            for (var j = 0; j < imageInfoArrayLength; j++) {
-                var image = images[j];
-                var imageInfo = imageInfoArray[j];
-                imagesInfo[image] = imageInfo;
-            }
-            return imagesInfo;
-        });
+    return final;
 }
 
-function getMaterials(mtlPath, hasMaterialGroups) {
-    if (hasMaterialGroups && defined(mtlPath)) {
-        return Material.parse(mtlPath);
+function removeEmptyMeshes(meshes) {
+    var final = [];
+    var meshesLength = meshes.length;
+    for (var i = 0; i < meshesLength; ++i) {
+        var mesh = meshes[i];
+        mesh.primitives = removeEmptyPrimitives(mesh.primitives);
```

Comment:
Similar comment to this and `removeEmptyMesages`, you can use a combination of filter and map here to make the code easier to read (and smaller).